homeassistant/components/home_connect/binary_sensor.py

```diff
@@ -119,19 +120,61 @@ async def async_setup_entry(
     async_add_entities: AddEntitiesCallback,
 ) -> None:
     """Set up the Home Connect binary sensor."""
+    added_entities: set[str | None] = set()
 
-    entities: list[BinarySensorEntity] = []
-    for appliance in entry.runtime_data.data.values():
+    def get_entities_for_appliance(
+        appliance: HomeConnectApplianceData,
+    ) -> list[BinarySensorEntity]:
+        remove_listener: Callable[[], None] | None = None
+
+        def handle_removed_device() -> None:
+            """Handle removed device."""
+            for entity_unique_id in added_entities.copy():
+                if entity_unique_id and appliance.info.ha_id in entity_unique_id:
+                    added_entities.remove(entity_unique_id)
+            assert remove_listener
+            remove_listener()
```

Comment:
why do we create one listener per entity we create?

Reply:
Is not per entity, it is per appliance.

But now that I think about that, we could save the appliance ids, and when the listener is called, compare the appliances we got and the appliances we have, and remove the ones that are no longer present. (just like the example at IQS rules 😅)